src/pool/pool_jemalloc.c

```diff
@@ -0,0 +1,423 @@
+// universally opted out of by setting the hook pointers to NULL, or selectively opted out of
+// by returning failure. Note that once the extent hook is set, the structure is accessed directly
+// by the associated arenas, so it must remain valid for the entire lifetime of the arenas.
+// (from https://jemalloc.net/jemalloc.3.html)
+static extent_hooks_t arena_extent_hooks = {
+    .alloc = arena_extent_alloc,
+    .dalloc = arena_extent_dalloc,
+    .destroy = arena_extent_destroy,
+    .commit = arena_extent_commit,
+    .decommit = arena_extent_decommit,
+    .purge_lazy = arena_extent_purge_lazy,
+    .purge_forced = arena_extent_purge_forced,
+    .split = arena_extent_split,
+    .merge = arena_extent_merge,
+};
+
+static void *je_malloc(void *pool, size_t size) {
+    assert(pool);
+    jemalloc_memory_pool_t *je_pool = (jemalloc_memory_pool_t *)pool;
+    // MALLOCX_TCACHE_NONE is set, because jemalloc can mix objects from different arenas inside
+    // the tcache, so we wouldn't be able to guarantee isolation of different providers.
+    int flags = MALLOCX_ARENA(je_pool->arena_index) | MALLOCX_TCACHE_NONE;
+    void *ptr = mallocx(size, flags);
+    if (ptr == NULL) {
+        TLS_last_allocation_error = UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+        return NULL;
+    }
+
+    return ptr;
+}
+
+static umf_result_t je_free(void *pool, void *ptr) {
+    (void)pool; // unused
+    assert(pool);
+
+    if (ptr != NULL) {
+        dallocx(ptr, MALLOCX_TCACHE_NONE);
+    }
+
+    return UMF_RESULT_SUCCESS;
+}
+
+static void *je_calloc(void *pool, size_t num, size_t size) {
+    assert(pool);
+    size_t csize = num * size;
+    void *ptr = je_malloc(pool, csize);
+    if (ptr == NULL) {
+        // TLS_last_allocation_error is set by je_malloc()
+        return NULL;
+    }
+
+    memset(ptr, 0, csize); // TODO: device memory is not accessible by host
+    return ptr;
+}
+
+static void *je_realloc(void *pool, void *ptr, size_t size) {
+    assert(pool);
+    if (size == 0 && ptr != NULL) {
+        dallocx(ptr, MALLOCX_TCACHE_NONE);
+        TLS_last_allocation_error = UMF_RESULT_SUCCESS;
+        return NULL;
+    } else if (ptr == NULL) {
+        return je_malloc(pool, size);
+    }
+
+    jemalloc_memory_pool_t *je_pool = (jemalloc_memory_pool_t *)pool;
+    // MALLOCX_TCACHE_NONE is set, because jemalloc can mix objects from different arenas inside
+    // the tcache, so we wouldn't be able to guarantee isolation of different providers.
+    int flags = MALLOCX_ARENA(je_pool->arena_index) | MALLOCX_TCACHE_NONE;
+    void *new_ptr = rallocx(ptr, size, flags);
+    if (new_ptr == NULL) {
+        TLS_last_allocation_error = UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+        return NULL;
+    }
+
+    return new_ptr;
+}
+
+static void *je_aligned_alloc(void *pool, size_t size, size_t alignment) {
+    assert(pool);
+    jemalloc_memory_pool_t *je_pool = (jemalloc_memory_pool_t *)pool;
```

Comment:
I see that this code is repeated 3 times - could you make a function get_flags(je_pool) and move it there?